File: modules/native-audio-session/android/src/main/java/expo/modules/nativeaudiosession/NativeAudioSessionModule.kt
package expo.modules.nativeaudiosession

import android.content.Context
import android.media.AudioManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise

class NativeAudioSessionModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw IllegalStateException("React context is not available")

  private val audioManager: AudioManager
    get() = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager

  override fun definition() = ModuleDefinition {
    Name("NativeAudioSession")

    AsyncFunction("getRoute") { promise: Promise ->
      try {
        val am = audioManager
        val route = when {
          am.isBluetoothA2dpOn || am.isBluetoothScoOn -> "bluetooth"
          am.isWiredHeadsetOn -> "wired"
          am.isSpeakerphoneOn -> "speaker"
          else -> "earpiece"
        }
        promise.resolve(route)
      } catch (e: Exception) {
        promise.reject("AudioSessionError", "Failed to get audio route: ${e.message}", e)
      }
    }

    AsyncFunction("setRoute") { route: String, promise: Promise ->
      try {
        val am = audioManager
        am.mode = AudioManager.MODE_IN_COMMUNICATION

        when (route) {
          "speaker" -> {
            am.isSpeakerphoneOn = true
          }
          "earpiece" -> {
            am.isSpeakerphoneOn = false
          }
          else -> {
            promise.reject("AudioSessionError", "Invalid route: $route", null)
            return@AsyncFunction
          }
        }
        promise.resolve(null)
      } catch (e: Exception) {
        promise.reject("AudioSessionError", "Failed to set audio route: ${e.message}", e)
      }
    }
  }
}

File: modules/native-audio-session/ios/NativeAudioSessionModule.swift
import ExpoModulesCore
import AVFoundation

public class NativeAudioSessionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NativeAudioSession")

    AsyncFunction("getRoute") { () -> String in
      let session = AVAudioSession.sharedInstance()
      let currentRoute = session.currentRoute

      for output in currentRoute.outputs {
        switch output.portType {
        case .builtInSpeaker:
          return "speaker"
        case .builtInReceiver:
          return "earpiece"
        case .bluetoothA2DP, .bluetoothLE, .bluetoothHFP:
          return "bluetooth"
        case .headphones, .headsetMic:
          return "wired"
        default:
          continue
        }
      }
      return "unknown"
    }

    AsyncFunction("setRoute") { (route: String) in
      let session = AVAudioSession.sharedInstance()

      do {
        try session.setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker, .allowBluetooth])

        if route == "speaker" {
          try session.overrideOutputAudioPort(.speaker)
        } else {
          try session.overrideOutputAudioPort(.none)
        }

        try session.setActive(true)
      } catch {
        throw Exception(name: "AudioSessionError", description: "Failed to set audio route: \(error.localizedDescription)")
      }
    }
  }
}
